README
Vinty - manage OpenSSL certificates for use with OpenVPN.
===========================================================

See the AUTHORS file for the list of authors.

License: GNU GENERAL PUBLIC LICENSE (see the LICENSE file for details)


0. Download
-----------
::

    $ git clone http://code.google.com/p/vinty/


1. Included applications
   ---------------------

   - vinty       -- commandline tool to create/renew/revoke certificates.
                    See vinty(8) for usage information.

   - vintypasswd -- set or remove passphrases from RSA keys generated with
                    vinty. See vintyapasswd(1) for usage information.

2. Web Application
   ---------------

   A web application to manage your certificates is included and has all
   the functionality of the vinty commandline utility.

   Additionally, certificate/key pairs, together with the CA's certificate
   can be downloaded as a zipfile from the web application.

   Another feature of the web application is the ability to publish the
   CRL from the application, by calling an external script.

3. Configuration
   -------------

   The main configuration file is /etc/vinty/vinty.cfg. The default
   configuration should be edited before either the commandline tool
   or the web application are used.

4. Bugs
   ----

   The vinty commandline application and the webfrontend cannot
   easily work together. See vinty(8) for details.

5. Make debian package
   -------------------
   cd $DIR/vinty
   dch (command to edit the changelog if needed)
   dpkg-buildpackage


debian/changelog
vinty (0.2.1) unstable; urgency=medium

  * silly mistake in 0.2

 -- Frank Gevaerts <frank.gevaerts@fks.be>  Tue, 14 Jan 2014 18:42:09 +0100

vinty (0.2) unstable; urgency=medium

  * Compatibility with newer openssl
  * Use sudo for interacting with the rest of the system.

 -- Frank Gevaerts <frank.gevaerts@fks.be>  Tue, 14 Jan 2014 18:15:40 +0100

vinty (0.1) unstable; urgency=low

  * Prepare for initial release

 -- Jeroen Hooyberghs <jeroen.hooyberghs@fks.be>  Fri, 07 Oct 2011 10:39:40 +0100


vinty.py
#!/usr/bin/python

## Revision history ############################################################
## Revision history ############################################################
__author__  = 'Wouter Eerdekens <info@fks.be>'
__date__    = '2011-08-12'
__version__ = '0.1'
__history__ = """
2011-08-12 - Prepare for initial release <jeroen.hooyberghs@fks.be>
2006-07-26 - initial version.
"""
## Imports #####################################################################
import os, sys
from os import path
from optparse import OptionParser
import pki.cmd, pki.util

## Functions ###################################################################
def error(msg):
   sys.stderr.write('Error: ' + msg + '\n')
   sys.exit(1)

# ------------------------------------------------------------------------------
def parse_argv(argv):
   cn      = ''
   usage   = """%prog --init
       %prog --create client|server CN
       %prog --renew client|server CN
       %prog --revoke CN
       %prog --list valid|revoked|all"""
   version = '%prog ' +  __version__
   parser  = OptionParser(usage=usage, version=version)
   parser.add_option('--init', action='store_true', default=False,
                     help='initialise the PKI infrastructure')
   parser.add_option('--list', metavar='TYPE', type='choice',
                     choices=('valid', 'revoked', 'all'),
                     help='list all certificates of TYPE')
   parser.add_option('--create', metavar='TYPE', type='choice',
                     choices=(pki.CERT_CLIENT, pki.CERT_SERVER),
                     help='create a certificate/key pair of TYPE')
   parser.add_option('--renew', metavar='TYPE', type='choice',
                     choices=(pki.CERT_CLIENT, pki.CERT_SERVER),
                     help='renew a certificate of TYPE')
   parser.add_option('--revoke', action='store_true', default=False,
                     help='revoke a certificate')
   opt, arg = parser.parse_args(argv)
   if opt.init:
      if opt.create or opt.renew or opt.revoke or opt.list:
         parser.error('--init cannot be combined with other options')
      if len(arg) != 0:
         parser.error('--init doesn\'t take an argument')
   elif opt.list:
      if opt.init or opt.create or opt.renew or opt.revoke:
         parser.error('--list cannot be combined with other options')
      if len(arg) != 0:
         parser.error('--list doesn\'t take extra arguments')
   elif opt.create:
      if opt.init or opt.list or opt.renew or opt.revoke:
         parser.error('--create cannot be combined with other options')
      if len(arg) == 1:
         cn = arg[0]
      else:
         parser.error('--create requires you to specify a CN')
   elif opt.renew:
      if opt.init or opt.create or opt.revoke or opt.list:
         parser.error('--renew cannot be combined with other options')
      if len(arg) == 1:
         cn = arg[0]
      else:
         parser.error('--renew requires you to specify a CN')
   elif opt.revoke:
      if opt.init or opt.create or opt.renew or opt.list:
         parser.error('--revoke cannot be combined with other options')
      if len(arg) == 1:
         cn = arg[0]
      else:
         parser.error('--revoke requires you to specify a CN')
   else:
      parser.print_help()
      sys.exit(1)
   return opt, cn

# ------------------------------------------------------------------------------
def info(msg):
   sys.stdout.write(msg)
   sys.stdout.flush()

# ------------------------------------------------------------------------------
def pki_init(config):
   sys.stdout.write('Creating PKI infrastructure: ')
   sys.stdout.flush()
   result, msg = pki.cmd.create_pki(config)
   if result:
      print 'ok'
   else:
      print 'failed'
      error(msg)

# ------------------------------------------------------------------------------
def cert_create(config, cn):
   type = opt.create
   cn   = pki.util.strip_invalid(cn)
   if cn:
      info('Creating ' + type + ' certificate with CN ' + cn + ': ')
      result, msg = pki.cmd.create_cert(config, cn, type)
      if result:
         print 'ok (serial: ' + pki.util.serial_from_cn(config, cn) + ')'
      else:
         print 'failed'
         error(msg)
   else:
      error('invalid CN given')

# ------------------------------------------------------------------------------
def cert_renew(config, cn):
   type = opt.renew
   cn   = pki.util.strip_invalid(cn)
   if cn:
      serial = pki.util.serial_from_cn(config, cn)
      if serial:
         info('Renewing ' + type + ' certificate with CN ' + cn + ': ')
         result, msg = pki.cmd.renew_cert(config, cn, opt.renew)
         if result:
            print ' ok (new serial: ' + pki.util.serial_from_cn(config,cn) + ')'
         else:
            print ' failed'
            error(msg)
      else:
         error('no serial found for ' + cn)
   else:
      error('invalid CN given')

# ------------------------------------------------------------------------------
def cert_revoke(config, cn):
   cn = pki.util.strip_invalid(cn)
   if cn:
      serial = pki.util.serial_from_cn(config, cn)
      if serial:
         info('Revoking certificate with CN ' + cn + ', serial ' + serial +': ')
         result, msg = pki.cmd.revoke_cert(config, cn, opt.revoke)
         if result:
            print ' ok'
         else:
            print ' failed'
            error(msg)
      else:
         error('no serial found for ' + cn)
   else:
      error('invalid CN given')

## We're being called as a script ##############################################
if __name__ == '__main__':
   opt, cn = parse_argv(sys.argv[1:])
   config  = pki.util.parse_config()
   if not config:
      error('unable to read/validate configuration')
   if opt.init:
      pki_init(config)
   else:
      if not pki.cmd.check_pki(config):
         error('PKI infrastructure not valid (first run with --init)')
      if opt.list:
         pki.cmd.list_certs(config, opt.list)
      elif opt.create:
         cert_create(config, cn)
      elif opt.renew:
         cert_renew(config, cn)
      elif opt.revoke:
         cert_revoke(config, cn)


scripts/publish_crl.sh
#!/bin/bash

## Revision history ############################################################
# author  : Wouter Eerdekens <info@fks.be>
# date    : 2011-08-12
# version : 0.1
# history : 2011-08-12 - Prepare for initial release <jeroen.hooyberghs@fks.be>
#           2006-07-26 - initial version.
################################################################################

## Read configuration.
[ ! -f /etc/vinty/scripts.conf ] && echo "error: cannot read config" && exit 1
. /etc/vinty/scripts.conf

## Read commandline arguments
CRL="$1"
[ ! -f "$CRL" ] && echo "error: cannot read CRL from $CRL" && exit 1

## Publish the CRL
if [ -z "$SSH_SERVER" ]
then
   if [ -n "$COPY_USER" -a "$(id -un)" != "$COPY_USER" ]
   then
      sudo -u "$COPY_USER" "$0" "$CRL"
      RV=$?
   else
      if cp "$CRL" "$OPENVPN_DIR"
      then
         $OPENVPN_CMD
         RV=$?
      else
         RV=1
      fi
   fi
else
   if scp -P "$SSH_SERVER_PORT" -i "$SSH_SERVER_KEY" \
      "$KEY_DIR/$CRL" "$SSH_SERVER_LOGIN@$SSH_SERVER:$OPENVPN_DIR"
   then
      ssh -p "$SSH_SERVER_PORT" -i "$SSH_SERVER_KEY" \
          "$SSH_SERVER_LOGIN@$SSH_SERVER" $OPENVPN_CMD
      RV=$?
   else
      RV=1
   fi
fi
exit $RV


vintypasswd.py
#!/usr/bin/python

## Revision history ############################################################
__author__  = 'Wouter Eerdekens <info@fks.be>'
__date__    = '2011-08-12'
__version__ = 0.1
__history__ = """
2011-08-12 - Prepare for initial release <jeroen.hooyberghs@fks.be>
2006-07-26 - initial version.
"""
################################################################################

## Imports #####################################################################
from getpass import getpass
from optparse import OptionParser
import commands, os, os.path, re, sys, tempfile
import pki, pki.util

## Functions ###################################################################
def error(msg):
   sys.stderr.write('Error: ' + msg + '\n')
   sys.exit(1)

# ------------------------------------------------------------------------------
def parse_argv(argv):
   keyfile = ''
   usage   = """%prog --set KEY
       %prog --remove KEY"""
   version = '%prog ' + __version__
   parser  = OptionParser(usage=usage, version=version)
   parser.add_option('--set', action='store_true', default=False,
                     metavar="KEY", help='add a passphrase to KEY')
   parser.add_option('--remove', action='store_true', default=False,
                     metavar="KEY", help='remove the passphrase from KEY')
   opt, arg = parser.parse_args(argv)
   if opt.set and opt.remove:
      parser.error('--set and --remove are mutually exclusive')
   elif opt.set or opt.remove:
      if len(arg) == 1:
         keyfile = arg[0]
      else:
         parser.error('no keyfile given')
   else:
      parser.print_help()
      sys.exit(1)
   return opt, keyfile

# ------------------------------------------------------------------------------
def remove_passphrase(config, keyfile, remove_only=True):
   if remove_only:
      passwd = getpass()
   else:
      passwd = getpass('Old password: ')
   tmpfile = keyfile + '.tmp'
   try:
      fd, fname = tempfile.mkstemp()
      os.write(fd, passwd)
      os.close(fd)
   except:
      sys.stderr.write('Problem creating secure temporary file...\n')
      sys.exit(1)
   else:
      cmd = pki.OPENSSL_COMMANDS['del_passphrase'] \
            % { 'openssl' : config.get('openssl', 'path'),
                'pwfile'  : fname,
                'inkey'   : keyfile,
                'outkey'  : tmpfile
              }
      status, output = commands.getstatusoutput(cmd)
      os.remove(fname)
      if status:
         sys.stderr.write('Problem removing old passphrase:\n' + output)
         sys.exit(1)
      else:
         try:
            os.remove(keyfile)
            os.rename(tmpfile, keyfile)
         except:
            sys.stderr.write('Problem storing new keyfile.\n')

# ------------------------------------------------------------------------------
def set_passphrase(config, keyfile):
   if pki.util.is_encrypted(keyfile):
      remove_passphrase(config, keyfile, remove_only=False)
   pw1     = getpass()
   pw2     = getpass('Retype password: ')
   tmpfile = keyfile + '.tmp'
   if pw1 != pw2:
      sys.stderr.write('Password mismatch.\n')
      sys.exit(1)
   else:
      try:
         fd, fname = tempfile.mkstemp()
         os.write(fd, pw1)
         os.close(fd)
      except:
         sys.stderr.write('Problem creating secure temporary file...\n')
         sys.exit(1)
      else:
         cmd = pki.OPENSSL_COMMANDS['set_passphrase'] \
               % { 'openssl' : config.get('openssl', 'path'),
                   'pwfile'  : fname,
                   'inkey'   : keyfile,
                   'outkey'  : tmpfile
                 }
         status, output = commands.getstatusoutput(cmd)
         os.remove(fname)
         if status:
            sys.stderr.write('Problem setting passphrase:\n' + output)
            sys.exit(1)
         else:
            try:
               os.remove(keyfile)
               os.rename(tmpfile, keyfile)
            except:
               sys.stderr.write('Problem storing new keyfile.\n')

## We're being called as a script ##############################################
if __name__ == '__main__':
   opt, keyfile = parse_argv(sys.argv[1:])
   config  = pki.util.parse_config()
   if not os.path.isfile(keyfile):
      error('Unable to open ' + keyfile + '.\n')
   else:
      if pki.util.is_private_key(keyfile):
         if opt.set:
            print 'Setting passphrase on ' + keyfile
            set_passphrase(config, keyfile)
         else:
            print 'Removing passphrase from ' + keyfile
            if pki.util.is_encrypted(keyfile):
               remove_passphrase(config, keyfile)
            else:
               error(keyfile + ' has no passphrase set.')
      else:
         error(keyfile + ' is not a private key.')
